modules/pool/lib/pool-onchain-data.ts

```diff
@@ -0,0 +1,211 @@
+import { PrismaPoolType } from '@prisma/client';
+import { BigNumber, formatFixed } from '@ethersproject/bignumber';
+import ElementPoolAbi from '../abi/ConvergentCurvePool.json';
+import LinearPoolAbi from '../abi/LinearPool.json';
+import LiquidityBootstrappingPoolAbi from '../abi/LiquidityBootstrappingPool.json';
+import ComposableStablePoolAbi from '../abi/ComposableStablePool.json';
+import GyroEV2Abi from '../abi/GyroEV2.json';
+import VaultAbi from '../abi/Vault.json';
+import aTokenRateProvider from '../abi/StaticATokenRateProvider.json';
+import WeightedPoolAbi from '../abi/WeightedPool.json';
+import StablePoolAbi from '../abi/StablePool.json';
+import MetaStablePoolAbi from '../abi/MetaStablePool.json';
+import StablePhantomPoolAbi from '../abi/StablePhantomPool.json';
+import { JsonFragment } from '@ethersproject/abi';
+
+const abi: JsonFragment[] = Object.values(
+    // Remove duplicate entries using their names
+    Object.fromEntries(
+        [
+            ...ElementPoolAbi,
+            ...LinearPoolAbi,
+            ...LiquidityBootstrappingPoolAbi,
+            ...ComposableStablePoolAbi,
+            ...GyroEV2Abi,
+            ...VaultAbi,
+            ...aTokenRateProvider,
+            ...WeightedPoolAbi,
+            ...StablePoolAbi,
+            ...StablePhantomPoolAbi,
+            ...MetaStablePoolAbi,
+            ...ComposableStablePoolAbi,
+            //...WeightedPoolV2Abi,
+        ].map((row) => [row.name, row]),
+    ),
+);
+
+const getSwapFeeFn = (type: string) => {
+    if (type === 'ELEMENT') {
+        return 'percentFee';
+    } else if (type === 'FX') {
+        return 'protocolPercentFee';
+    } else {
+        return 'getSwapFeePercentage';
+    }
+};
+
+interface PoolInput {
+    id: string;
+    address: string;
+    type: PrismaPoolType;
+    tokens: {
+        address: string,
+        token: {
+            decimals: number,
+        }
+    }[];
+    version?: number;
+}
+
+interface OnchainData {
+    poolTokens: [string[], BigNumber[]];
+    totalSupply: BigNumber;
+    virtualSupply?: BigNumber;
+    actualSupply?: BigNumber;
+    swapFee: BigNumber;
+    swapEnabled?: boolean;
+    protocolYieldFeePercentageCache?: BigNumber;
+    rate?: BigNumber;
+    weights?: BigNumber[];
+    targets?: [BigNumber, BigNumber];
+    wrappedTokenRate?: BigNumber;
+    amp?: [BigNumber, boolean, BigNumber];
+    tokenRates?: [BigNumber, BigNumber];
+    tokenRate?: BigNumber[];
+    metaPriceRateCache?: [BigNumber, BigNumber, BigNumber][];
+}
+
+const defaultCalls = (
+    { id, address, type }: PoolInput,
+    vaultAddress: string,
+    multicaller: Multicaller3
+) => {
+    multicaller.call(`${id}.poolTokens`, vaultAddress, 'getPoolTokens', [id]);
+    multicaller.call(`${id}.totalSupply`, address, 'totalSupply');
+    multicaller.call(`${id}.virtualSupply`, address, 'getVirtualSupply');
+    multicaller.call(`${id}.actualSupply`, address, 'getActualSupply');
```

Comment:
I would move these to the pool type specific queries, otherwise its an extra 2 queries per pool that will either fail or are not needed.

Reply:
that's the part where we need to know the composable stable pool type, which depends on adding logic based on the factory address. i understand we don't need overfetching, but we fetch so much already that the cost is marginal. i tested both versions (this vs previous) and there is no perfomance impact.